src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+        }
+        catch (Exception ex)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            var error = await HandleErrorDuringSubscribeAsync(message, ex);
+            await SendErrorResultAsync(message, error);
+        }
+    }
+
+    /// <summary>
+    /// Creates an <see cref="ExecutionError"/> for an unknown <see cref="Exception"/>.
+    /// </summary>
+    protected virtual Task<ExecutionError> HandleErrorDuringSubscribeAsync(OperationMessage message, Exception ex)
+        => Task.FromResult<ExecutionError>(new UnhandledError("Unable to set up subscription for the requested field.", ex));
+
+    /// <summary>
+    /// Sends a single result to the client for a subscription or request, along with a notice
+    /// that it was the last result in the event stream.
+    /// </summary>
+    protected virtual async Task SendSingleResultAsync(OperationMessage message, ExecutionResult result)
+    {
+        await SendDataAsync(message.Id!, result);
+        await SendCompletedAsync(message.Id!);
+    }
+
+    /// <summary>
+    /// Sends an execution error to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionError executionError)
+        => SendErrorResultAsync(message.Id!, executionError);
+
+    /// <summary>
+    /// Sends an execution error to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(string id, ExecutionError executionError)
+        => SendErrorResultAsync(id, new ExecutionResult { Errors = new ExecutionErrors { executionError } });
+
+    /// <summary>
+    /// Sends an error result to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionResult result)
+        => SendErrorResultAsync(message.Id!, result);
+
+    /// <summary>
+    /// Sends an error result to the client during set-up of a subscription.
+    /// </summary>
+    protected abstract Task SendErrorResultAsync(string id, ExecutionResult result);
+
+    /// <summary>
+    /// Sends a data packet to the client for a subscription event.
+    /// </summary>
+    protected abstract Task SendDataAsync(string id, ExecutionResult result);
+
+    /// <summary>
+    /// Sends a notice that a subscription has completed and no more data packets will be sent.
+    /// </summary>
+    protected abstract Task SendCompletedAsync(string id);
+
+    /// <summary>
+    /// Executes a GraphQL request. The request is inside <see cref="OperationMessage.Payload"/>
+    /// and will need to be deserialized by <see cref="IGraphQLSerializer.ReadNode{T}(object?)"/>
+    /// into a <see cref="GraphQLRequest"/> instance.
+    /// </summary>
+    protected abstract Task<ExecutionResult> ExecuteRequestAsync(OperationMessage message);
```

Comment:
Both implementations are identical and code copy-pasted 😕 . Do you intend to keep it as is to keep implementations simple and independent?